L1Trigger/TrackerTFP/src/MiniHoughTransform.cc

```diff
@@ -0,0 +1,285 @@
+      for (int k = 0; k < numCells_; k++)
+        swap(tmp[k], stubsCells[(channel / numCells_) * numBinsInv2R_ + channel % numCells_ + k * numCells_]);
+      slb(tmp, streamsSLB[channel], lost[channel]);
+    }
+    // dynamic load balancing stage 1
+    vector<vector<StubMHT*>> streamsDLB(numBinsInv2R_);
+    for (int node = 0; node < numNodes_; node++) {
+      vector<vector<StubMHT*>> tmp(numChannel_);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsSLB[(node / numCells_) * numNodes_ + node % numCells_ + k * numCells_]);
+      dlb(tmp);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsDLB[node * numChannel_ + k]);
+    }
+    // dynamic load balancing stage 2
+    vector<vector<StubMHT*>> streamsMHT(numBinsInv2R_);
+    for (int node = 0; node < numNodes_; node++) {
+      vector<vector<StubMHT*>> tmp(numChannel_);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsDLB[node + k * numNodes_]);
+      dlb(tmp);
+      for (int k = 0; k < numChannel_; k++)
+        swap(tmp[k], streamsMHT[node * numChannel_ + k]);
+    }
+    // fill output product
+    for (int channel = 0; channel < numBinsInv2R_; channel++) {
+      const vector<StubMHT*>& stubs = streamsMHT[channel];
+      TTDTC::Stream& stream = accepted[region_ * numBinsInv2R_ + channel];
+      stream.reserve(stubs.size());
+      for (StubMHT* stub : stubs)
+        stream.emplace_back(stub ? stub->frame() : TTDTC::Frame());
+    }
+  }
+
+  // perform finer pattern recognition per track
+  void MiniHoughTransform::fill(const vector<StubHT*>& stubs, vector<deque<StubMHT*>>& streams, int channel) {
+    if (stubs.empty())
+      return;
+    int id;
+    auto differentHT = [&id](StubHT* stub){ return id != stub->trackId(); };
+    auto differentMHT = [&id](StubMHT* stub){ return !stub || id != stub->trackId(); };
+    for (auto it = stubs.begin(); it != stubs.end();) {
+      const auto start = it;
+      id = (*it)->trackId();
+      it = find_if(it, stubs.end(), differentHT);
+      const int size = distance(start, it);
+      // create finer track candidates stub container
+      vector<vector<StubMHT*>> mhtCells(numCells_);
+      for (vector<StubMHT*>& mhtCell : mhtCells)
+        mhtCell.reserve(size);
+      // fill finer track candidates stub container
+      for (auto stub = start; stub != it; stub++) {
+        const double r = (*stub)->r();
+        const double chi = (*stub)->phi();
+        // identify finer track candidates for this stub
+        // 0 and 1 belong to bigger sub inv2R; 0 and 2 belong to smaller sub track phi
+        vector<int> cells;
+        cells.reserve(numCells_);
+        const bool compA = 2. * abs(chi) < phiT_.base();
+        const bool compB = 2. * abs(chi) < abs(r * inv2R_.base());
+        const bool compAB = compA && compB;
+        if (chi >= 0. && r >= 0.) { cells.push_back(3); if (compA) cells.push_back(1); if(compAB) cells.push_back(2); }
```

Comment:
This code seems to assume that mhtNumBinsInv2R = mhtNumBinsPhiT = 2. Please throw fatal error if this is not the case.